rust/headless-client/src/ipc_service.rs

```diff
@@ -193,9 +198,9 @@ enum HandlerOk {
     ServiceTerminating,
 }
 
-impl Handler {
-    async fn new(server: &mut IpcServer) -> Result<Self> {
-        dns_control::deactivate()?;
+impl<'a> Handler<'a> {
+    async fn new(server: &mut IpcServer, dns_controller: &'a mut DnsController) -> Result<Self> {
+        dns_controller.deactivate()?;
```

Comment:
Given this, is the default of `may_be_active: true` redundant?

Reply:
No, maybe `active` is the wrong word for it. It's more like a `dirty` flag in a cache, it means we need to deactivate our control if we should not be in control.

If I split up the I/O benchmarking from the deterministic parts as I wrote in the other thread, I might be able to remove this completely.

There's 3 reasons to deactivate DNS control:
1. The IPC service or Headless Client just started, and we need to recover from a possible Firezone crash or system crash
2. Connlib initiated a disconnect and we need to bail out of this iteration of the IPC service's main loop
3. The GUI initiated a sign-out and we need to stop controlling DNS

Number 1 should always happen. Numbers 2 and 3 can afford to be slow since Firezone is shutting down anyway. So I think this optimization made sense a few days ago when it removed the defensive "deactivate during sign-in" but now it may be unnecessary.

Reply:
> The IPC service or Headless Client just started, and we need to recover from a possible Firezone crash or system crash

Can we move this concern to the `main` of the respective binary?